src/cis2/mod.rs

```diff
@@ -0,0 +1,352 @@
+//! This module contains types and functions for interacting with smart
+//! contracts following the [CIS-2](https://proposals.concordium.software/CIS/cis-2.html) specification.
+//!
+//! The type [Cis2Contract](Cis2Contract) act as a wrapper around the
+//! [Client](crate::endpoints::Client) and a contract address providing
+//! functions for querying and making transactions to smart contract.
+mod types;
+
+use crate::{common, endpoints::Client, id, types as sdk_types};
+use sdk_types::{smart_contracts, transactions, ContractAddress};
+use smart_contracts::concordium_contracts_common;
+use std::convert::{From, TryFrom};
+use thiserror::*;
+pub use types::*;
+
+/// A wrapper around the client representing a CIS2 token smart contract and
+/// provides functions for interaction.
+///
+/// Note that cloning is cheap and is therefore the intended way of sharing
+/// between multiple tasks.
+#[derive(Debug, Clone)]
+pub struct Cis2Contract<'a> {
+    client:        Client,
+    address:       ContractAddress,
+    contract_name: concordium_contracts_common::ContractName<'a>,
```

Comment:
I think using a shared reference is going to cause people pain. For example they can't easily spawn off a client into a background task since that'll require `'static: 'a`.

I recommend either just `OwnedContractName` or `Arc<OwnedContractName>`.

Reply:
I'll go with `Arc<OwnedContractName>` then, since cloning should be cheaper